### PR TITLE
Dmsayo/190/push notifications

### PR DESCRIPTION
**Pull request by dmsayo**
Add push notifications. Currently tested on desktop web browsers only (unable to test mobile until deployment), we can keep an eye out for any bugs and make the appropriate tickets.

**Also for deployment a collection in the database named "subscriptions" may have to be made in order to properly store subscription objects.**

### **Gifs**

ezgif com-gif-maker (1)

ezgif com-gif-maker (2)

### CODE REVIEW

File: chat/fe/src/services/PushService.ts [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+            applicationServerKey: urlB64ToUint8Array(applicationServerPublicKey),
+        })
+        if (subscribed) {
+            saveSubscriptionOnServer(subscribed);
+        }
+    } else {
+        const currentSubscription = await serviceWorker.pushManager.getSubscription();
+        const unsubscribed = (await serviceWorker.pushManager.getSubscription()).unsubscribe();
+        if (unsubscribed) {
+            removeSubscriptionOnServer(currentSubscription);
```
Comment (adamjberg) [inline chat/fe/src/services/PushService.ts:49]: This should be awaited

File: chat/be/package.json [hunk 2]
```diff
@@ -8,6 +8,7 @@
     "dotenv": "^10.0.0",
```
Comment (adamjberg) [inline chat/be/package.json:8]: It looks like you ran npm install at some point.  Please remove chat/be/package-lock.json and make sure to use yarn in the future.

**Review by adamjberg — CHANGES_REQUESTED** [review]
Few requests across the files.  Can you add a folder for chat/docs and then start a `.md` file that explains how to get vapid keys? I will probably need to generate new ones for production, which is where the .env request comes in to play

File: chat/be/server.js [hunk 3]
```diff
@@ -197,9 +207,29 @@ async function run() {
...
       { $inc: { unreadCount: 1 } }
     );
-
+    
     io.emit('message', newMessage);
 
+    const currentRoom = await Room.findOne({
+      _id: mongodb.ObjectId(id)
+    });
+    const subscriptions = await Subscriptions.find({ user: { $in: currentRoom.users, $ne: mongodb.ObjectId(req.user) } }).toArray();
```
Comment (adamjberg) [inline chat/be/server.js:216]: Can we change the model name to PushNotification?  Susbcription could evnetually be used for something like billing subscription.

Comment (dmsayo) [reply chat/be/server.js:214]: I realized I haven't changed the database model name when I pushed the changes at 4, I'll fix it on Tuesday.

**Review by adamjberg — APPROVED** [review]

**Pull request merged**